main/metrics.c

```diff
@@ -0,0 +1,94 @@
+    METRICS_PREFIX "heap_free_bytes{capability=\"MALLOC_CAP_INTERNAL\"} %zu\n"
+    METRICS_PREFIX "heap_free_bytes{capability=\"MALLOC_CAP_DEFAULT\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_EXEC\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_32BIT\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_8BIT\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_DMA\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID2\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID3\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID4\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID5\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID6\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_PID7\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_SPIRAM\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_INTERNAL\"} %zu\n"
+    METRICS_PREFIX "heap_largest_free_block_bytes{capability=\"MALLOC_CAP_DEFAULT\"} %zu\n",
+    gw_metrics.received_advertisements,
+    esp_timer_get_time(),
+    get_total_free_bytes(MALLOC_CAP_EXEC),
+    get_total_free_bytes(MALLOC_CAP_32BIT),
+    get_total_free_bytes(MALLOC_CAP_8BIT),
+    get_total_free_bytes(MALLOC_CAP_DMA),
+    get_total_free_bytes(MALLOC_CAP_PID2),
+    get_total_free_bytes(MALLOC_CAP_PID3),
+    get_total_free_bytes(MALLOC_CAP_PID4),
+    get_total_free_bytes(MALLOC_CAP_PID5),
+    get_total_free_bytes(MALLOC_CAP_PID6),
+    get_total_free_bytes(MALLOC_CAP_PID7),
+    get_total_free_bytes(MALLOC_CAP_SPIRAM),
+    get_total_free_bytes(MALLOC_CAP_INTERNAL),
+    get_total_free_bytes(MALLOC_CAP_DEFAULT),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_EXEC),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_32BIT),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_8BIT),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_DMA),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_PID2),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_PID3),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_PID4),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_PID5),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_PID6),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_PID7),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_SPIRAM),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_INTERNAL),
+    (size_t) heap_caps_get_largest_free_block(MALLOC_CAP_DEFAULT));
+}
+
+char* ruuvi_get_metrics()
+{
+    char *buf = malloc(BUF_LEN * sizeof(char));
```

Comment:
I think this is a bit dangerous as I can't see where `buf` is freed. Maybe allocate `buf` at callee and pass the pointer + size as parameter?

Reply:
The buf is freed in http_server.c (see the other PR) in the same way as ruuvi.json is freed there after copying it into the http response

Reply:
Ok, @TheSomeMan has final say here

Reply:
it's commonly used approach in this project, so I think it's ok for now, but in future we can refactor it.